### PR TITLE
Implements hashing as a blanket trait for instances of `CanonicalSerialize`

### PR DESCRIPTION
**Pull request by huitseeker**
- this saves an alllocation w.r.t the issue-suggested approach by implementing `io::Write` on the input instance of `digest::Digest`,
- note that most instances of `digest::Digest` [already](https://gist.github.com/huitseeker/e827161413063e347ce5a496b66ff287) have an [`io::Write` instance](https://github.com/rustcrypto/hashes#hashing-readable-objects), but `CanonicalSerialize` consuming its `io::Write` argument prevents its usage,
- this hence implements `io::Write` on a [cheap newtype wrapper](https://rust-unofficial.github.io/patterns/patterns/behavioural/newtype.html)

closes: #86 

---

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: serialize/src/lib.rs [hunk 1]
```diff
@@ -94,6 +96,37 @@ pub trait CanonicalSerialize {
...
+    fn hash<H: Digest>(&self) -> GenericArray<u8, <H as Digest>::OutputSize> {
+        let mut hasher = H::new();
+        self.serialize(HashMarshaller(&mut hasher))
+            .expect("HashMarshaller::flush should be infaillible!");
+        hasher.finalize()
+    }
+}
+
+/// Hash is a (blanket) extension trait of CanonicalSerialize
+impl<T: CanonicalSerialize> Hash for T {}
```
Comment (Pratyush) [inline serialize/src/lib.rs:128]: Should we just make this a standalone method(s)?
```rust
fn hash_with_digest_compressed<D: Digest, T: CanonicalSerialize>() -> GenericArray<u8, <H as Digest>::OutputSize> { ... }

fn hash_with_digest_uncompressed<D: Digest, T: CanonicalSerialize>() -> GenericArray<u8, <H as Digest>::OutputSize> { ... }
```

Not sure of the trade-off between importing a trait (`Hash`) vs. two methods. What do you think?

Regardless, I think we should rename it from `Hash` to something else, as it collides with `core::hash::Hash`.

Comment (huitseeker) [reply serialize/src/lib.rs]: > Not sure of the trade-off between importing a trait (Hash) vs. two methods.

The [extension trait pattern](https://rust-lang.github.io/rfcs/0445-extension-trait-conventions.html) is clear enough that most will prefer it to manually importing several methods, I think.

I renamed the trait according to best practices.

Comment (huitseeker) [reply serialize/src/lib.rs]: I could also simply add those two methods to the `CanonicalSerialize` trait? Wdyt?

File: serialize/src/lib.rs [hunk 2]
```diff
@@ -94,6 +96,44 @@ pub trait CanonicalSerialize {
...
+
+    #[inline]
+    fn flush(&mut self) -> ark_std::io::Result<()> {
+        Ok(())
+    }
+}
+
+/// The CanonicalSerialize induces a natural way to hash the
+/// corresponding value, of which this is the convenience trait.
+pub trait CanonicalSerializeExt: CanonicalSerialize {
```
Comment (Pratyush) [inline serialize/src/lib.rs:118]: Maybe we call this something like `CanonicalSerializeThenHash`?

File: serialize/src/lib.rs (hunk 1 above)
Comment (Pratyush) [reply serialize/src/lib.rs]: I prefer the separation of concerns, and maybe we should put this behind a separate feature?

**Comment by Pratyush** [comment]
> but CanonicalSerialize consuming its io::Write argument prevents its usage,

Is it not possible to work around this by doing `&mut writer` (`&mut W` is `Write` whenever `W` is)

File: serialize/src/lib.rs (hunk 2 above)
Comment (huitseeker) [reply serialize/src/lib.rs:118]: - The serialization is never materialized anywhere, whereas `CanonicalSerializeThenHash` implies that it is, in the way the peudo-code in #86 did.
- The best practice for an extension trait is really to finish in `Ext`. `CanonicalSerializeHashExt`?

**Comment by huitseeker** [comment]
@Pratyush Sure, though this signature change would change the API of `CanonicalSerialize` in a backwards-incompatible way. 

Moreover, I'm not sure we want to depend on the `std::io::Write` implementations in RustCrypto/hashes: 
- the `Digest` trait does not guarantee its existence on its own, 
- it's implemented from unrelated (MAC) trait impls through a macro [which future in the code base isn't assured](https://github.com/RustCrypto/traits/pull/217/files)
- the no-std context adds an additional wrinkle.

File: serialize/src/lib.rs (hunk 2 above)
Comment (Pratyush) [reply serialize/src/lib.rs:118]: `CanonicalSerializeHashExt` sounds good.

**Comment by Pratyush** [comment]
Ok that makes sense, let's proceed as is.

**Review by Pratyush — APPROVED** [review]

**Comment by Pratyush** [comment]
Oh @huitseeker if you could add a CHANGELOG entry that would be great (also for #263)

**Review by ValarDragon — COMMENTED** [review]
Thanks for adding this!

**Pull request merged**